README.md
# fraction
Transformed simple fractions into beautifully animated results

This program runs best on PC since the graphic package graphics.py depveloped by John Zelle is optimized for PC

orbit2.py contains codes for generating patterns from fractions

fraction.py contains codes for the start menu interface


history.txt
pattern0

domino.py
#domino.py
#Jessy Quint and Khanh Nghiem
#Final Project
#December 14,2016
#This program generates a domino-like image to represent the user-inputted fraction.
#The program will also take in mouse clicks that cause the domino pips to change color.

from graphics import *
from buttonclass import *

def drawCover(gwin):
    """Drawing the covers to avoid overlay"""
    cover=Rectangle(Point(5-1.25,6.7-1.25),Point(8+1.25,6.7+1.25))
    cover.draw(gwin)
    cover.setFill("white")
    cover.setOutline("white")

    cover2=Rectangle(Point(5-1.25,2.5-1.25),Point(8+1.25,2.5+1.5))
    cover2.draw(gwin)
    cover2.setFill("white")
    cover2.setOutline("white")

def __inputBox(gwin,height):
    """ __inputBox(gwin,height) is a hidden construction method
    that creates an input box"""        
    inputBox=Entry(Point(2,height),2)
    inputBox.setSize(60)
    inputBox.setFill("white")
    inputBox.draw(gwin)
    return inputBox

def inputBox(gwin):
    """ inputBox(gwin) calls the hidden constructor
    method so that we can create two input boxes in the same step """
    numInputBox = __inputBox(gwin,6.7)
    denInputBox = __inputBox(gwin,2.7)
    return numInputBox,denInputBox

class Dot:
    """ DieView is a widget that displays a graphical
    representation of a standard six-sided die."""
    
    def __init__(self, win, center, size):
        """Create a view of a die, e.g.:
           d1 = DieView(myWin, Point(40,50), 20)
        creates a die centered at (40,50) having sides
        of length 20."""

        # first define some standard values
        self.win = win
        self.background = "white" # color of die face
        self.foreground = "black" # color of the pips
        self.psize = 0.1 * size   # radius of each pip
        hsize = size / 2.0        # half of size
        offset = 0.6 * hsize      # distance from center to outer pips

        # create a square for the face
        cx, cy = center.getX(), center.getY()
        p1 = Point(cx-hsize, cy-hsize)
        p2 = Point(cx+hsize, cy+hsize)

        # Create 7 circles for standard pip locations
        self.pips = [ self.__makePip(cx-offset, cy-offset), #bottom-left 0
                      self.__makePip(cx-offset, cy), #middle-left 1
                      self.__makePip(cx-offset, cy+offset), #top-left 2
                      self.__makePip(cx, cy), #center 3
                      self.__makePip(cx+offset, cy-offset), #bottom-right 4
                      self.__makePip(cx+offset, cy), #middle-right 5
                      self.__makePip(cx+offset, cy+offset), #top-right 6

                      self.__makePip(cx,cy+offset), #top-middle 7
                      self.__makePip(cx,cy-offset)] #bottom-middle 8

        # Create a table for which pips are on for each value
        # note: 10 will be created by two 5s
        self.onTable = [ [], [3], [1,5], [7,4,0], 
            [0,2,4,6], [0,2,3,4,6], [2,7,6,0,8,4], [2,7,6,0,8,4,3], [0,1,2,4,5,6,7,8], [0,1,2,3,4,5,6,7,8] ]

    def __makePip(self, x, y):
        """Internal helper method to draw a pip at (x,y)"""
        pip = Circle(Point(x,y), self.psize)
        pip.setFill(self.background)
        pip.setOutline(self.background)
        pip.draw(self.win)
        return pip


    def setValue(self, value):
        """ Set this die to display value."""
        # Turn all the pips off
        for pip in self.pips:
            pip.setFill(self.background)

        # Turn the appropriate pips back on
        for i in self.onTable[value]:
            self.pips[i].setFill(self.foreground)
            

class ColorDot(Dot): #subclass of the superclass 'DieView'

    def setColor(self, color): #method to change color of pips on die drawn
        self.foreground = color #setValue() in DieView uses self.foreground
                                #to set pip color and in this line we set that color

def domino(win):

    #buttons
    visualize=Button(win,1,2,Point(5,9),"Visualize","misty rose")
    quitButton=Button(win,0.5,0.5,Point(9.5,9.5),"X","light gray")

    #draw input boxes to get the fraction
    numInput,denInput=inputBox(win)
    
    fracText=Text(Point(2,5.5),"__")
    fracText.draw(win)
    fracText.setSize(60)
    
    pt=win.getMouse()
    error = None
    
    #while not quit
    while not quitButton.isClicked(pt):
        if visualize.isClicked(pt):
            drawCover(win)
            if error:
                error.undraw()
            try:
                #draw input boxes
                num,den=int(numInput.getText()),int(denInput.getText())

                if num not in [1,2,3,4,5,6,7,8,9,10] or den not in [1,2,3,4,5,6,7,8,9,10]:
                    kill=10/0 #this will create an error and jump to the except line
                
                #create boolean flags to later change pip color
                twoNum = False
                twoDen = False

                #draw pips according to input values
                if num==10:
                    Ndot1 = ColorDot(win, Point(5.8,6.7), 2.5)
                    Ndot1.setColor("dark blue")
                    Ndot1.setValue(5)
                    Ndot2 = ColorDot(win, Point(8,6.7), 2.5)
                    Ndot2.setColor("dark blue")
                    Ndot2.setValue(5)
                    twoNum = True
                else:
                    numDots = ColorDot(win, Point(7,6.7), 2.5)
                    numDots.setColor("dark blue")
                    numDots.setValue(num)
                    
                if den==10:
                    Ddot1 = ColorDot(win, Point(5.8,2.5), 2.5)
                    Ddot1.setColor("dark blue")
                    Ddot1.setValue(5)
                    Ddot2 = ColorDot(win, Point(8,2.5), 2.5)
                    Ddot2.setColor("dark blue")
                    Ddot2.setValue(5)
                    twoDen = True
                else:    
                    denDots = ColorDot(win, Point(7,3), 2.5)
                    denDots.setColor("dark blue")
                    denDots.setValue(den)
                
                dotFrac=Text(Point(7,5.5),"_____")
                dotFrac.draw(win)
                dotFrac.setSize(60)
            except:
                error=Text(Point(5,1),"Please enter whole numbers smaller than 11")
                error.draw(win)
            
        elif not visualize.isClicked(pt): #if a new click was made and it was not on quit button
            try:
                col = color_rgb(randrange(0,220),randrange(0,220),randrange(0,220))
                #change pip colors
                if twoNum:
                    Ndot1.setColor(col)
                    Ndot1.setValue(5)
                    Ndot2.setColor(col)
                    Ndot2.setValue(5)
                else:
                    numDots.setColor(col)
                    numDots.setValue(num)
                    
                if twoDen:
                    Ddot1.setColor(col)
                    Ddot1.setValue(5)
                    Ddot2.setColor(col)
                    Ddot2.setValue(5)
                else:
                    denDots.setColor(col)
                    denDots.setValue(den)
            except:
                print("")
                
        pt=win.getMouse()
    win.close()



fraction.py
#fraction.py
#Khanh Nghiem and Jessy Quint
#Fiinal Project
#December 14, 2016

#This module holds main function that runs the entire program by
#importing other modules holding classes and methods necessary
#for computations. This module also draws opening graphics and menu
#used to begin program.
#IMPORTANT NOTE: This program is designed to be run from a PC computer.
#The program could crash if run on a Mac device.

from graphics import *
from buttonclass import *
from orbit2 import *
from domino import *

def __inputBox(gwin,height,word):
    """ __inputBox(gwin,height,word) is a hidden construction method
    that creates an input box with a prompt for the user and draws both
    into a graphic window """        
    prompt=Text(Point(155,height),word)
    prompt.setSize(10)
    prompt.draw(gwin)
    inputBox=Entry(Point(450,height),2)
    inputBox.setSize(60)
    inputBox.setFill("white")
    inputBox.draw(gwin)
    return inputBox

def inputBox(gwin,prt1,prt2):
    """ inputBox(gwin,prompt1,prompt2) calls the hidden constructor
    method so that we can create two input boxes in the same step """
    numInputBox = __inputBox(gwin,300, prt1)
    line=Text(Point(450,345),"____")
    line.draw(gwin)
    line.setSize(40)
    denInputBox = __inputBox(gwin,450, prt2)
    return numInputBox,denInputBox

def main():
    win=GraphWin("Fraction",600,600)
    win.setBackground("white")

    #title
    title=Text(Point(300,70),"Fun with Fractions")
    title.setSize(30)
    title.setFill("midnight blue")
    title.draw(win)

    #add cute cat image objects
    meow=Image(Point(115,70),"meow.png")
    meow.draw(win)
    meow2=Image(Point(485,70),"meow.png")
    meow2.draw(win)

    #draw buttons
    quitButton=Button(win,20,20,Point(565,25),"X","light gray")
    menu=Button(win,30,450,Point(300,150),"MENU","lavender")
    Domino=Button(win,45,200,Point(175,200),"Domino","misty rose")
    Orbit=Button(win,45,200,Point(425,200),"Orbit","LightCyan2")

    #outline boxes
    DpromptBox=Rectangle(Point(20,270),Point(580,330))
    DpromptBox.setOutline("misty rose")
    DpromptBox.setFill("misty rose")
    DpromptBox.draw(win)
    OpromptBox=Rectangle(Point(20,340),Point(580,460))
    OpromptBox.setOutline("LightCyan2")
    OpromptBox.setFill("LightCyan2")
    OpromptBox.draw(win)

    DpromptBox2=Rectangle(Point(23,273),Point(577,327))
    DpromptBox2.setOutline("white")
    DpromptBox2.setFill("white")
    DpromptBox2.draw(win)
    OpromptBox2=Rectangle(Point(23,343),Point(577,457))
    OpromptBox2.setOutline("white")
    OpromptBox2.setFill("white")
    OpromptBox2.draw(win)

    #description text for user
    Dprompt=Text(Point(300,300),"DOMINO will produce a visual representation of an inputted fraction, similar to a domino.\nClick anywhere within the blank space in the window to change the color of the domino.")
    Dprompt.setSize(10)
    Dprompt.draw(win)
    Oprompt=Text(Point(300,400),"ORBIT will create a pattern that represents an inputted fraction or fractions.\nThere will be two circles of the same radius, and a point on each circle (point A, point B)\nmoving along the circumference at relative speeds corresponding to the fraction.\nBy drawing a horizontal line from point A and a vertical line from point B,\nwe trace the intersection point--thus creating a beautiful pattern.")
    Oprompt.setSize(10)
    Oprompt.draw(win)
    
    pt=win.getMouse()
    
    while not quitButton.isClicked(pt):
        if Orbit.isClicked(pt):
            #get text from input boxes
            #create graphic window to display orbit animation
            win2=GraphWin("Orbit",750,750)
            win2.setBackground("white")
            #call orbit funtion from other module to run program
            #and pass in the window, numerator and denominator values to use
            orbit2(win2)
            pt=win.getMouse()


        elif Domino.isClicked(pt):
            #do same as above but call domino() function
            win2 = GraphWin("Domino",450,450)
            win2.setCoords(0,0,10,10)
            win2.setBackground("white")
            domino(win2)
            pt=win.getMouse()
                    
        else:
            pt=win.getMouse()
    win.close()
main()


buttonclass.py
from graphics import *
from random import *

class Button:

    """A rectangular is a labeled rectangle in a window. It is enabled or
disabled with the activate() and deactivated() methods. The isClicked(pt)
method returns True if the button is enabled and pt is inside it."""
    
    #first set up the construtor method
    #this is automatically called each time button class is instantiated

    def __init__(self,win,height,width,centerPt,words,color):
        """Creates a rectangular button, where:
        win is the GraphWin obect where the button will be drawn,
        height is an integer
        width is an integer
        centerPt is a Point object the button will be centered on
        wrds is a string that will appear on the button"""
        
        x,y=centerPt.getX(), centerPt.getY()
        self.xmax=x+width/2
        self.xmin=x-width/2
        self.ymax=y+height/2
        self.ymin=y-height/2
        self.rect=Rectangle(Point(self.xmin,self.ymin),Point(self.xmax,self.ymax))
        self.rect.draw(win)
        self.rect.setFill(color)
        self.rect.setOutline(color)
        self.label=Text(centerPt,words)
        self.label.draw(win)
        self.activate()

    def isClicked(self,pt):
        """returns true if Point p is inside"""
        return (self.active and
                self.xmin <= pt.getX() <= self.xmax and
                self.ymin <= pt.getY() <= self.ymax)

    def activate(self):
        """Sets this button to active."""
        self.active=True
        self.rect.setWidth(2)
        self.label.setFill('black')

    def deactivate(self):
        """Sets this button to inactive mode"""
        self.active=False
        self.rect.setWidth(1)
        self.label.setFill('gray')

    def setLabel(self,newLabel):
        """mutator method"""
        self.label.setText(newLabel)

    def getLabel(self):
        return self.label.getText()

    def undraw(self):
        self.rect.undraw()
        self.label.undraw()
        
def main():
    print("We're in the Button Class")
    win=GraphWin("Dice Roller",200,200)
    rollButton=Button(win,20,120,Point(100,105),"Roll Dice")
    quitButton=Button(win,20,40,Point(100,180),"Quit")
    quitButton.deactivate()

    die1=DieView(win, Point(50,50), 50)
    die2=DieView(win, Point(150,50), 50)
    
    pt=win.getMouse()
    while not rollButton.isClicked(pt):
        pt=win.getMouse()
    
    quitButton.activate()
    value1=randrange(1,7)
    value2=randrange(1,7)
    die1.setValue(value1)
    die2.setValue(value2)

    pt=win.getMouse()
    while not quitButton.isClicked(pt):
        value1=randrange(1,7)
        value2=randrange(1,7)
        die1.setValue(value1)
        die2.setValue(value2)
        pt=win.getMouse()
    win.close()
if __name__=='__main__':
    main() 
    


orbit2.py
from graphics import *
from math import *
from time import *
from turtle import *
from buttonclass import *
from random import *

def __inputBox(gwin,height):
    """ __inputBox(gwin,height) is a hidden construction method
    that creates an input box"""        

    inputBox=Entry(Point(225,height),2)
    inputBox.setSize(60)
    inputBox.setFill("white")
    inputBox.draw(gwin)
    return inputBox

def inputBox(gwin):
    """ inputBox(gwin) calls the hidden constructor
    method so that we can create two input boxes in the same step """
    numInputBox = __inputBox(gwin,425)
    denInputBox = __inputBox(gwin,575)
    return numInputBox,denInputBox

class DancingPoint():
    def __init__(self,pointA,pointB,r,num,den,scale,sleeptime): #num is short for numerator
        #All points are tuples of x and y not objects
        #Later we'll access the tuples to get x and y to draw point objects
        self.num=num
        self.den=den
        self.pointA=pointA
        self.pointB=pointB
        self.pointAs=[pointA] #start each list that holds tuples with the starting points
        self.pointBs=[pointB]
        self.pointCs=[]
        self.maxrad=-pi*1/(8*scale) #the biggest movement (in radians) that we can make
        self.r=r #radius
        self.scale=scale #scaling the number of points according to user input to prevent too few or too many points
        self.sleeptime=sleeptime #sleeptime will change to prevent drawing too slow or too fast
        self.DPList=[] #this list will hold the Point objects corresponding to the Point C tuples

    def createPoint(self):
        """ this method will create all the points needed as tuples"""
        #radA and radB adjust the unit movement on circle A, cirle B 
        self.radA=self.maxrad/self.den
        self.radB=self.maxrad/self.num

        """DXA,DYA and DXB,DYB are lists of all unit movements to finish one cycle around the circumference of circle A, circle B in pixels
        these values are returned from the hidden method __createPoint
        multiply DXA,DYA,DXB,DYB by both num and den values so that loops are completed with the same number of movements"""
        
        self.DXA,self.DYA=self.__createPoint(self.den,self.pointAs,self.radA)
        self.DXA=self.DXA*self.den*self.num #multiply by den completes 1 cycle, then multiply by num to return to the starting point
        self.DYA=self.DYA*self.num*self.den #multiply by num completes 1 cycle, then multiply by den to return to the starting point
        
        self.DXB,self.DYB=self.__createPoint(self.num,self.pointBs,self.radB)
        self.DXB=self.DXB*self.den*self.num
        self.DYB=self.DYB*self.num*self.den
        
        self.__createDancingPoint() #create all point Cs

    def __createPoint(self,value,pointList,rad):
        """ __createPoint is the hidden method that actually creates all points on a circle. This method will be called separately for each circle."""
        DX=[]
        DY=[]
        numPoints=16*value*self.scale #number of points on a circle
        for i in range(numPoints):
            pointi=pointList[0] #getting the starting point of the circle
            pointiX=pointi[0]
            pointiY=pointi[1]

            rad1=rad+rad*i-pi*1/2 #how far away point i is from the starting point of the circle

            #dx, dy represent the movement in each direction to get to point i
            dx=(self.r+self.r*sin(rad1))-(self.r+self.r*sin(rad+rad*(i-1)-pi*1/2))
            dy=(self.r*cos(rad1))-(self.r*cos(rad+rad*(i-1)-pi*1/2))

            #adding all movements to two lists, DX, DY
            DX.append(-dx)
            DY.append(-dy)

            #adding point i to the list of all points on a circle
            pointi1X=pointiX-(self.r+self.r*sin(rad1))
            pointi1Y=pointiY-(self.r*cos(rad1))
            pointList.append((pointi1X,pointi1Y))
    
        pointList.pop(0) #delete the starting point because we already have it as the last point of the list

        return DX,DY

    def __createDancingPoint(self):
        """ __createDancingPoint is the hidden method that actually creates all points that create the final pattern"""

        #make sure there exists the same number of pointAs and pointBs to create pointCs of intersection
        self.pointAs=self.pointAs*self.num
        self.pointBs=self.pointBs*self.den
        numPoints=16*self.num*self.den*self.scale
        for i in range (numPoints): #create pointCs by using corresponding pointAs and pointBs
            pointA=self.pointAs[i]
            pointB=self.pointBs[i]
            pointC=(pointB[0],pointA[1]) #pointC is created by taking the y-value of pointA and the x-value of corresponding pointB
            self.pointCs.append(pointC)
        
    
    def draw(self,win):
        """ this method does all the visualization"""

        #draw legs to map intersection
        self.legA=Line(Point(150,225),Point(750,225))
        self.legA.draw(win)
        self.legB=Line(Point(600,600),Point(600,75))
        self.legB.draw(win)
        
        colA = "gray"
        colB = "gray"

        #set counts at start to 0
        countA=0
        countB=0

        #keep track of how many times each circle completes one cycle around the circumference
        self.countAText=Text(Point(225,100),"0")
        self.countAText.draw(win)
        self.countAText.setSize(24)
        
        self.countBText=Text(Point(400,525),"0")
        self.countBText.draw(win)
        self.countBText.setSize(24)

        #do all the drawing
        for i in range (len(self.pointAs)): #can use self.pointBs or self.pointCs bc they are of the same length
            pointi=self.pointAs[i]
            pointiX=pointi[0]
            pointiY=pointi[1]
            pointiDraw=Point(pointiX,pointiY) #draw point A

            self.legA.move(self.DXA[i],self.DYA[i]) #move horizontal line according to point A
            
            line=Line(pointiDraw,Point(225,225)) #draw line from center of circle A to point A
            line.draw(win)
            
            if i!=0 and i % (len(self.pointAs)/self.num) == 0: #if circle A finishes a cycle, change the color
                countA+=1
                self.countAText.setText(countA)
                if colA == "white":
                    colA = "gray"
                else:
                    colA = "white"
            line.setFill(colA)

            #do same for point Bs
            pointi=self.pointBs[i]
            pointiX=pointi[0]
            pointiY=pointi[1]
            pointiDraw=Point(pointiX,pointiY)

            self.legB.move(self.DXB[i],self.DYB[i])
            
            line=Line(pointiDraw,Point(525,525))
            line.draw(win)

            if i!=0 and i % (len(self.pointBs)/self.den) == 0:
                countB+=1
                self.countBText.setText(countB)
                if colB == "white":
                    colB = "gray"
                else:
                    colB = "white"
            line.setFill(colB)

            #do same for points Cs
            pointi=self.pointCs[i]
            pointiX=pointi[0]
            pointiY=pointi[1]
            pointiDraw=Point(pointiX,pointiY)
            pointiDraw.draw(win)
            pointiDraw.setFill("purple")

            #keep track of all Point objects for later uses
            self.DPList.append(pointiDraw)

            #connects the points of the patterns with lines
            if i>0:
                line=Line(self.DPList[i],self.DPList[i-1])
                line.draw(win)
                line.setFill("dark blue")
                
            else: #connect the last point to the first point
                line=Line(self.DPList[i],Point(600,225))
                line.draw(win)
                line.setFill("dark blue")
                
            sleep(self.sleeptime)

        #update counts
        self.countAText.setText(countA+1)
        self.countBText.setText(countB+1)

    def savePicture(self,infile,outfile):
        """ This method allows user to save the pattern they created to an image file"""
        img=Image(Point(0,0),infile) #get the source file to manipulate

        #loop through the list of all point Cs        
        for i in range(len(self.pointCs)-1):
            #get point i
            point1=self.pointCs[i]
            x1=round(point1[0])-375
            y1=round(point1[1])-75

            #get point i+1
            point2=self.pointCs[i+1]
            x2=round(point2[0])-375
            y2=round(point2[1])-75

            #drawing all the points in between point i and point i+1 (using parametric line)
            for i in range(20):
                t=0.05*i
                xp=round(x1+t*(x2-x1))
                yp=round(y1+t*(y2-y1))
                img.setPixel(xp,yp,"darkblue")             

        #connecting the last point to the first point
        point1=self.pointCs[0]
        x1=round(point1[0])-375
        y1=round(point1[1])-75

        point2=self.pointCs[len(self.pointCs)-1]
        x2=round(point2[0])-375
        y2=round(point2[1])-75

        for i in range (50):
            t=0.02*i
            xp=round(x1+t*(x2-x1))
            yp=round(y1+t*(y2-y1))
            img.setPixel(xp,yp,"darkblue") 

        #export the processed image
        img.save(outfile)

    def reset(self,gwin):
        """This method allows for another run. We call this method before starting a new iteration of the program."""
        self.countBText.undraw()
        self.countAText.undraw()
        self.legA.undraw()
        self.legB.undraw()
        coverCirc1=Circle(Point(225,225),75)
        coverCirc1.setFill("white")
        coverCirc1.draw(gwin)
        coverCirc2=Circle(Point(525,525),75)
        coverCirc2.setFill("white")
        coverCirc2.draw(gwin)

    def reset2(self,gwin):
        """This is an alternate version of the reset method. We call this method when user hit reset."""
        self.legA.undraw()
        self.legB.undraw()
        self.legA.draw(gwin)
        self.legB.draw(gwin)
        self.countBText.setText("0")
        self.countAText.setText("0")
        coverCirc1=Circle(Point(225,225),75)
        coverCirc1.setFill("white")
        coverCirc1.draw(gwin)
        coverCirc2=Circle(Point(525,525),75)
        coverCirc2.setFill("white")
        coverCirc2.draw(gwin)
           
def orbit2(win):
    #draw quit button
    quitButton=Button(win,20,20,Point(725,25),"X","light gray")
    
    #before user can run the program, they need to decide whether to save what they create 
    prompt=Text(Point(375,100),"This pattern was created by running this program.\nDo you want to save the pattern that you create as an image file?")
    prompt.draw(win)
    
    yes=Button(win,40,75,Point(300,500),"Yes","lavenderblush")
    no=Button(win,40,75,Point(400,500),"No","lavenderblush")

    example=Image(Point(350,300),"example.gif")
    example.draw(win)

    pt=win.getMouse()

    while not yes.isClicked(pt) and not no.isClicked(pt) and not quitButton.isClicked(pt):
        pt=win.getMouse()
        
    if yes.isClicked(pt):
        saveFile=True
        prompt.undraw()
        yes.undraw()
        no.undraw()
        example.undraw()

    elif no.isClicked(pt):
        saveFile=False
        prompt.undraw()
        yes.undraw()
        no.undraw()
        example.undraw()

    elif quitButton.isClicked(pt):
        win.close()
    if not quitButton.isClicked(pt):
        #drawing all buttons
        startButton=Button(win,40,75,Point(375,60),"Start","lavenderblush")

        #if they don't want to save the pattern, they can reset the program as often as they choose
        if saveFile==False:
            resetButton=Button(win,40,75,Point(375,105),"Reset","lavenderblush")

        #configuring the circles
        r=75
        xA,yA=225,225
        xB,yB=525,525

        #draw circles, starting points, and starting connecting lines
        circleA=Circle(Point(xA,yA),r)
        circleA.draw(win)
        
        circleB=Circle(Point(xB,yB),r)
        circleB.draw(win)
        
        pointA=Point(xA+r,yA)
        pointB=Point(xB+r,yB)

        pointC=Point(xB+r,yA)
        pointC.draw(win)
        
        lineA=Line(pointA,pointC)
        lineA.draw(win)
        lineB=Line(pointB,pointC)
        lineB.draw(win)

        line1=Line(Point(xA,yA),pointA)
        line1.draw(win)
        line2=Line(Point(xB,yB),pointB)
        line2.draw(win)


        #drawing fraction input boxes
        numInput,denInput=inputBox(win)
        fracText=Text(Point(xA,450),"__")
        fracText.draw(win)
        fracText.setSize(60)
        fracText.setStyle("bold")

        #open the history.txt file to see how many times the user have saved images
        file=open("history.txt","r")
        infile=file.read()
        file.close()

        #get the last time they saved an image, increase it by one
        i=infile.index("pattern")
        filenum=infile[i+7]
        filenum=int(filenum)+1

        error = None
        dancingPoint = None #indicates the user hasn't used the program

        #set the canvas and output image file name
        canvas="canvas.gif"
        imgname="pattern"+str(filenum)+".gif"

        #fracList keeps track of the fractions user has inputed in 1 run
        fracList=""
        fractionList=Text(Point(350,650),fracList)
        fractionList.draw(win)

        #while user still run the program
        while not quitButton.isClicked(pt):
            pt=win.getMouse()

            if startButton.isClicked(pt):
                if error: #if there is an error, undraw the error message
                    error.undraw()
                if dancingPoint: #if the user has run the program before, reset  
                    dancingPoint.reset(win)
                try:     
                    num,den=int(numInput.getText()),int(denInput.getText()) #get inputs

                    if num not in [1,2,3,4,5,6,7,8,9,10] or den not in [1,2,3,4,5,6,7,8,9,10]:
                        kill=10/0 #this creates an error, program jumps to the except line

                    #configure the the program accordingly to user inputs 
                    elif num<5 and den<5: #if small numbers
                        scale=2 #more points
                        sleeptime=0.01 #slower movements

                    elif num>5 and den>5: #if big numbers
                        scale=1 #fewer points
                        sleeptime=0 #faster movements

                    else: #if moderate inputs (1 big, 1 small)
                        scale=1 #fewer points
                        sleeptime=0.01 #slower movements

                    lineA.undraw() #undraw the starting connecting lines
                    lineB.undraw()

                    #create an instance of the Dancing Point class
                    dancingPoint=DancingPoint((xA+r,yA),(xB+r,yB),r,num,den,scale,sleeptime)
                    dancingPoint.createPoint()
                    dancingPoint.draw(win)

                    #if user wants to save pattern as an image file, call the .saveimage method
                    if saveFile==True:
                        dancingPoint.savePicture(canvas,imgname) #the first time, use the blank canvas
                        canvas=imgname #then keep using the exported file as the canvas to continue drawing until program ends
                        notice=Text(Point(350,675),"This pattern has been saved as "+imgname)
                        notice.draw(win)

                    #update fraction list
                    fracList=fracList+str(num)+"/"+str(den)+", "
                    fractionList.setText(fracList)
                    
                
                except:
                    error=Text(Point(350,685),"Please enter whole numbers smaller than 10")
                    error.setFill("VioletRed3")
                    error.draw(win)
        
            elif saveFile == False and resetButton.isClicked(pt):
                #cover the pattern
                cover=Rectangle(Point(450,150),Point(602,302))
                cover.setFill("white")
                cover.setOutline("white")
                cover.draw(win)
                
                #call the alternate version of reset
                dancingPoint.reset2(win)
                fracList=""
                fractionList.setText(fracList)
                numInput.setText("")
                denInput.setText("")

        #if the user saved an image file, update the history file        
        if saveFile==True:
            file=open("history.txt","w")
            content="pattern"+str(filenum)+"\n"+fracList+".\n"+infile
            file.write(content)
            file.close()
    
        win.close()
            
